### ISSUE TITLE
"rxjs-no-explicit-generics" warns against "new BehaviorSubject<string[]>([])"

### ISSUE THREAD

**Issue by stenzengel** [cartant/rxjs-tslint-rules#95]
The rule "rxjs-no-explicit-generics" warns against the following code (only for illustration):
```typescript
const bs = new BehaviorSubject<string[]>([]); //[tslint] Explicit generic type arguments are forbidden (rxjs-no-explicit-generics)
bs.next(["foo"]);
```

When the generic parameter is removed, the TypeScript compiler creates an error:
```typescript
const bs = new BehaviorSubject([]);
bs.next(["foo"]); // [ts] Type 'string' is not assignable to type 'never'.
```

My current way of fixing this is by removing the generic parameter and using a type assertion for the constructor parameter, but this notation does not look better than the first one:
```typescript
const bs = new BehaviorSubject([] as string[]);
bs.next(["foo"]);
```
I'm using TypeScript 3.2.4 and rxjs-tslint-rules 4.23.1.

This is not a big deal for me, but perhaps there is an another way to fix this?

**Comment by cartant** [comment]
I don't consider this to be something that needs to be fixed. IMO, explicitly declaring type parameters in situations in which they can be inferred is something to avoid. And I would rather keep that as a general rule than introduce exceptions for things like empty arrays. I'd favour taking the `as string[]` approach, but you could also use a TSLint flag to disable it for the next line:

```
/*tslint:disable-next-line:rxjs-no-explicit-generics*/
```

The intention of the rule is not to make things look better, but to avoid over-specifying types.

**Comment by cartant** [comment]
I think about it this way: here, we have only one type argument, but if we had a function that took several type arguments and wanted to pass an empty array literal to the function, if you wanted to specify the array type via a type argument, you would have to specify all of the type arguments - and that makes no sense. So, for me, it's best to avoid using type arguments to indicate types when `as` could be used instead (for situations in which types cannot be inferred).

**Comment by stenzengel** [comment]
Thanks for answer. I'm totally fine with it.

**Comment by criskrzysiu** [comment]
@cartant  ok lets say we have strict types enabled in ts and want to initialize subject with undefined but the type we want is string | undefined. We either have to disable the rule or use ugly undefined as string or prefix with comment to turn off tslint for this line.

**Comment by cartant** [comment]
> ... prefix with comment to turn off tslint for this line

Use a linter flag comment to disable it for that line. This issue might be closed, but there is an open PR - https://github.com/cartant/rxjs-tslint-rules/pull/99 - that relates to it and that PR contains some discussion. There are some subtleties and I don't anticipate changes being made to this rule anytime soon, as I have higher priorities.
